components/tidb_query_vec_executors/src/index_scan_executor.rs

```diff
@@ -344,148 +342,163 @@ impl IndexScanExecutorImpl {
-                let handle = self.decode_handle_from_key(key_payload)?;
+            // For normal index, it is placed at the end and any columns prior to it are
+            // ensured to be interested. For unique index, it is placed in the value.
+            DecodeIntHandle if key_payload.is_empty() => {
+                // This is a unique index, and we should look up PK int handle in the value.
+                let handle_val = self.decode_handle_from_value(value)?;
                 columns[self.columns_id_without_handle.len()]
                     .mut_decoded()
-                    .push_int(Some(handle));
+                    .push_int(Some(handle_val));
             }
-            // This is a unique index value, we should extract the int handle from the value.
             DecodeIntHandle => {
-                let handle = self.decode_handle_from_value(&value[value.len() - tail_len..])?;
+                // This is a normal index, and we should look up PK handle in the key.
+                let handle_val = self.decode_handle_from_key(key_payload)?;
                 columns[self.columns_id_without_handle.len()]
                     .mut_decoded()
-                    .push_int(Some(handle));
+                    .push_int(Some(handle_val));
             }
-            // This is a non-unique index value, we should extract the common handle from the key.
             DecodeCommonHandle => {
-                datum::skip_n(&mut key_payload, self.columns_id_without_handle.len())?;
+                // Otherwise, if the handle is common handle, we extract it from the key.
                 Self::extract_columns_from_datum_format(
                     &mut key_payload,
                     &mut columns[self.columns_id_without_handle.len()..],
                 )?;
             }
         }
+
         Ok(())
     }
 
-    // Process index values that are in old collation but don't contain common handles.
-    // NOTE: We should extract the index columns from the key first, and extract the handles from value if there is no handle in the key.
-    // Otherwise, extract the handles from the key.
-    fn process_old_collation_kv(
+    #[inline]
+    fn split_common_handle_from_index_value<'a>(value: &'a [u8]) -> Result<(&'a [u8], &'a [u8])> {
+        if value
+            .get(0)
+            .map_or(false, |c| *c == table::INDEX_VALUE_COMMON_HANDLE_FLAG)
+        {
+            let handle_len = (&value[1..]).read_u16().map_err(|_| {
+                other_err!(
+                    "Fail to read common handle's length from value: {}",
+                    hex::encode_upper(value)
+                )
+            })? as usize;
+            let handle_end_offset = 3 + handle_len;
+            if handle_end_offset > value.len() {
+                return Err(other_err!("`handle_len` is corrupted: {}", handle_len));
+            }
+            Ok(value[3..].split_at(handle_len))
+        } else {
+            Ok(value.split_at(0))
```

Comment:
Should we verify the length of common handle in this case?

Reply:
Seems no need to, because it's already done in `if handle_end_offset > value.len() {...}`